nixos/modules/services/backup/duplicity.nix

```diff
@@ -0,0 +1,99 @@
+{ config, lib, pkgs, ...}:
+
+with lib;
+
+let cfg = config.services.duplicity;
+
+in {
+  options.services.duplicity = {
+    enable = mkEnableOption "backups with duplicity";
+
+    root = mkOption {
+      type = types.path;
+      default = "/";
+      description = ''
+        Root directory to backup.
+      '';
+    };
+
+    include = mkOption {
+      type = types.listOf types.path;
+      default = [];
+      example = [ "/home" ];
+      description = ''
+        List of paths to include into the backups. See
+        <citerefentry><refentrytitle>duplicity</refentrytitle>
+        <manvolnum>1</manvolnum></citerefentry> for details on the syntax.
+      '';
+    };
+
+    exclude = mkOption {
+      type = types.listOf types.path;
+      default = [];
+      description = ''
+        List of paths to exclude from backups. See
+        <citerefentry><refentrytitle>duplicity</refentrytitle>
+        <manvolnum>1</manvolnum></citerefentry> for details on the syntax.
+      '';
+    };
+
+    targetUrl = mkOption {
+      type = types.str;
+      example = "s3://host:port/prefix";
+      description = ''
+        Target url to backup to. See
+        <citerefentry><refentrytitle>duplicity</refentrytitle>
+        <manvolnum>1</manvolnum></citerefentry> for supported urls.
+      '';
+    };
+
+    secretFile = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      description = ''
+        Path of a file containing secrets (gpg passphrase, access key...). It
+        is sourced before duplicity is run. For example, a file containing:
+          export PASSPHRASE='...'
+          export AWS_ACCESS_KEY_ID='...'
+          export AWS_SECRET_ACCESS_KEY='...'
+      '';
+    };
+
+    frequency = mkOption {
+      type = types.nullOr types.str;
+      default = "daily";
+      description = ''
+        Run duplicity with the given frequency (see
+        <citerefentry><refentrytitle>systemd.time</refentrytitle>
+        <manvolnum>7</manvolnum></citerefentry> for the format).
+        If null, do not run automatically.
+      '';
+    };
+
+    extraOptions = mkOption {
+      type = types.listOf types.str;
+      default = [];
+      example = [ "--full-if-older-than 1M" ];
+      description = ''
+        Extra options to pass to duplicity.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.duplicity = {
+      description = "Backup files with duplicity.";
+      script =
```

Comment:
Note that since you didn't set a user, this runs as root. Try looking into systemd's `ReadOnlyPaths` and `InaccessiblePaths` to limit this to a user so we don't need root. Probably even `DynamicUser` could work, then we don't need any fixed user at all.

Reply:
Thanks for the review.

I think that duplicity is expected to run as root, otherwise non world readable files would not be backed up.
What do you think?

I can still use other interesting options, such as StateDirectory.

Reply:
Ah yeah, good point. These options can still be applied when the user is root though, so even though it might need to run as root, we can lock it out of the things it shouldn't see (and deny it to write anything).